### ISSUE TITLE
[DataGrid] Columns not updating with `apiRef.current.updateColumns`

### ISSUE THREAD

**Issue by Janpot** [mui/mui-x#5970]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Using `apiRef.current.updateColumns` doesn't seem to update the existing columns.

### Expected behavior 🤔

It should update the columns

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/brave-breeze-quutgg?file=/demo.tsx

Steps:
1. click the "update columns" button
2. one grid will update (because it remounts when columns change). The other won't, which uses `updateColumns` to trigger an update.

### Context 🔦

Toolpad needs to update the columns of a grid in the canvas when a user makes changes in the editor. (currently we remount by regenerating a `key` when `columns` or `getRowId` changes)

### Your environment 🌎

https://codesandbox.io/s/brave-breeze-quutgg?file=/demo.tsx

**Comment by DanailH** [comment]
Hmm, interesting, I can reproduce it. The really interesting part is that the `updateColumns` does work, but not for all properties. For example, if you update the `width` that will work, but not the type.

**Comment by flaviendelangle** [comment]
The `type` is a special property because it causes a merge between the current column properties and the default column type properties.

I guess what is happening is that if you change the type in `updateColumns` it merges the column currently in state with the default properties of the new column types.
But the column currently in state was already merged with the default properties of the previous column types.
So there is no missing property and the new merge does basically nothing.

Not sure how to solve this one without adding a ton of complexity.

**Comment by Janpot** [comment]
> So there is no missing property and the new merge does basically nothing.

But there would be properties with new values, right? Or does the merge only add properties?

**Comment by flaviendelangle** [comment]
Small mistake on my side, we just don't apply the type default properties on update.

@Janpot yes, but the `type` property being a flag to tell to the grid "apply the default values of this column", it depends when we apply the default values and how we apply them.
For instance, if you pass `{ type: "date-time", field: "some field" }` to a column with a custom `sortComparator`, I don't think you want this update to override the custom `sortComparator` with the default one from `date-time`. 
You probably only want to apply the new default properties to properties that currently have their default values.

**Comment by Janpot** [comment]
If this is the intended behavior then I think it could be worth it to document it. Would it make sense for us to also offer a `setColumns` function to completely overwrite the columns? If we make it a `React.Dispatch<GridColumns>` then users can even create their own merging behavior.

**Comment by flaviendelangle** [comment]
> If this is the intended behavior then I think it could be worth it to document it.

Not what I'm saying
I'm just pointing out that the fix is probably non trivial and that we must be careful about the exact behavior.

> Would it make sense for us to also offer a setColumns function to completely overwrite the columns?

Why don't you use the `props.columns` if you want to update everything ?
We could add this method if it helps of course.

**Comment by Janpot** [comment]
> Why don't you use the props.columns if you want to update everything?

That doesn't work in my [codesandbox](https://codesandbox.io/s/brave-breeze-quutgg?file=/demo.tsx) and that seems intentional:

<img width="832" alt="Screenshot 2022-09-01 at 13 48 13" src="https://user-images.githubusercontent.com/2109932/187907419-98c73e65-c77d-459c-8797-d3ac0aa58c67.png">

**Comment by flaviendelangle** [comment]
Indeed
But even with a `setColumns` I guess that would be risky to update everything.

Another approach if the fix is too complicated is to add a param to `updateColumns` saying the grid to skip the merge and just set the new column to equal whats you passed to `updateColumns`

But let's try to do a clean fix before.

**Comment by bharatkashyap** [comment]
@flaviendelangle Should #5999 have closed this? If I understand correctly, it can't be closed yet.

**Comment by flaviendelangle** [comment]
Indeed, Github misunderstood my comment, good catch

**Comment by Janpot** [comment]
awesome, will try it out next week

**Comment by Janpot** [comment]
Not sure what happened. Either I didn't verify properly or it regressed again, but this issue is back:
codesandbox is broken for me so I recreated in a stackblitz: https://stackblitz.com/edit/react-hp7ypm-3dpbiy?file=Demo.tsx

Closing https://github.com/mui/mui-x/issues/9971 in favour of reopening this one

**Comment by cherniavskii** [comment]
@Janpot I believe the issue is caused by a typo:

```diff
-const COLUMNDEF1 = [{ field: 'foo', width: 200, type: 'datetime' }];
+const COLUMNDEF1 = [{ field: 'foo', width: 200, type: 'dateTime' }];
```

Fixed demo: https://stackblitz.com/edit/react-hp7ypm-8pb1hb?file=Demo.tsx

But it does expose an issue with the data grid - the `datetime` column type falls back to the `string` type, but it doesn't override the `valueFormatter` from the previous column type. I'll open a PR to fix this, thanks for the reproduction!

**Comment by Janpot** [comment]
Right, I was a bit sloppy porting the reproduction from https://github.com/mui/mui-x/issues/9971 back here. Can confirm that https://github.com/mui/mui-x/issues/9971 is fixed by https://github.com/mui/mui-x/pull/10041
